File: backend/controllers/ContactController.js
/**
 * Contact Controller
 * Single Responsibility: Handle contact form submissions
 */

const { v4: uuidv4 } = require('uuid');

class ContactController {
  constructor(collections, emailConfig) {
    this.collections = collections;
    this.emailConfig = emailConfig;
  }

  // POST /api/contact (public)
  submit = async (req, res) => {
    try {
      const { name, email, phone, message } = req.body;

      if (!name || !email || !message) {
        return res.status(400).json({
          success: false,
          message: 'Name, email, and message are required'
        });
      }

      const contact = {
        id: uuidv4(),
        name,
        email,
        phone: phone || '',
        message,
        is_read: false,
        notes: '',
        created_at: new Date().toISOString()
      };

      await this.collections.contactSubmissions.insertOne(contact);

      console.log(`✅ New contact submission from: ${email}`);

      // Send email notification (async, don't wait)
      this.emailConfig.sendContactNotification(name, email, phone || '', message)
        .catch(err => console.error('Email notification failed:', err.message));

      res.status(201).json({
        success: true,
        message: 'Thank you for your message. We will get back to you soon.',
        contact
      });
    } catch (error) {
      console.error('Submit contact error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to submit contact form'
      });
    }
  };

  // GET /api/admin/contacts (admin)
  list = async (req, res) => {
    try {
      const { unread_only } = req.query;
      const query = unread_only === 'true' ? { is_read: false } : {};

      const contacts = await this.collections.contactSubmissions
        .find(query, { projection: { _id: 0 } })
        .sort({ created_at: -1 })
        .toArray();

      res.json({
        success: true,
        contacts
      });
    } catch (error) {
      console.error('List contacts error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to list contacts'
      });
    }
  };

  // GET /api/admin/contacts/:id (admin)
  get = async (req, res) => {
    try {
      const { id } = req.params;

      const contact = await this.collections.contactSubmissions.findOne(
        { id },
        { projection: { _id: 0 } }
      );

      if (!contact) {
        return res.status(404).json({
          success: false,
          message: 'Contact not found'
        });
      }

      res.json({
        success: true,
        contact
      });
    } catch (error) {
      console.error('Get contact error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to get contact'
      });
    }
  };

  // PUT /api/admin/contacts/:id/read (admin)
  markAsRead = async (req, res) => {
    try {
      const { id } = req.params;

      const contact = await this.collections.contactSubmissions.findOne({ id });

      if (!contact) {
        return res.status(404).json({
          success: false,
          message: 'Contact not found'
        });
      }

      await this.collections.contactSubmissions.updateOne(
        { id },
        { $set: { is_read: true } }
      );

      res.json({
        success: true,
        message: 'Marked as read'
      });
    } catch (error) {
      console.error('Mark as read error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to mark as read'
      });
    }
  };

  // PUT /api/admin/contacts/:id/notes (admin)
  updateNotes = async (req, res) => {
    try {
      const { id } = req.params;
      const { notes } = req.body;

      const contact = await this.collections.contactSubmissions.findOne({ id });

      if (!contact) {
        return res.status(404).json({
          success: false,
          message: 'Contact not found'
        });
      }

      await this.collections.contactSubmissions.updateOne(
        { id },
        { $set: { notes: notes || '' } }
      );

      const updated = await this.collections.contactSubmissions.findOne(
        { id },
        { projection: { _id: 0 } }
      );

      res.json({
        success: true,
        contact: updated
      });
    } catch (error) {
      console.error('Update notes error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to update notes'
      });
    }
  };

  // DELETE /api/admin/contacts/:id (admin)
  delete = async (req, res) => {
    try {
      const { id } = req.params;

      const contact = await this.collections.contactSubmissions.findOne({ id });

      if (!contact) {
        return res.status(404).json({
          success: false,
          message: 'Contact not found'
        });
      }

      await this.collections.contactSubmissions.deleteOne({ id });

      console.log(`✅ Deleted contact: ${id}`);

      res.json({
        success: true,
        message: 'Contact deleted'
      });
    } catch (error) {
      console.error('Delete contact error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to delete contact'
      });
    }
  };
}

module.exports = ContactController;

File: backend/controllers/contact_controller.py
from fastapi import APIRouter, HTTPException, Depends, status, BackgroundTasks
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from models.schemas import ContactSubmission, ContactSubmissionCreate
from services.auth_service import auth_service
from services.email_service import email_service
from motor.motor_asyncio import AsyncIOMotorDatabase
from typing import List
from datetime import datetime, timezone
import uuid
import logging

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/contact", tags=["Contact"])
security = HTTPBearer(auto_error=False)


async def verify_admin(credentials: HTTPAuthorizationCredentials, db: AsyncIOMotorDatabase) -> dict:
    """Verify admin user from token"""
    if not credentials:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Not authenticated")
    
    payload = auth_service.decode_token(credentials.credentials)
    if not payload or payload.get("type") != "access":
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid token")
    
    user = await db.admin_users.find_one({"id": payload["sub"]}, {"_id": 0})
    if not user or not user.get("is_active", True):
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="User not found or disabled")
    
    return user


def create_contact_routes(db: AsyncIOMotorDatabase):
    """Create contact form routes"""
    
    @router.post("/", response_model=ContactSubmission, status_code=status.HTTP_201_CREATED)
    async def submit_contact(
        contact_data: ContactSubmissionCreate,
        background_tasks: BackgroundTasks
    ):
        """Submit a contact form (public endpoint)"""
        contact_id = str(uuid.uuid4())
        contact_doc = {
            "id": contact_id,
            **contact_data.model_dump(),
            "created_at": datetime.now(timezone.utc).isoformat(),
            "is_read": False,
            "notes": None
        }
        
        await db.contact_submissions.insert_one(contact_doc)
        logger.info(f"New contact submission from: {contact_data.email}")
        
        # Send email notification in background
        background_tasks.add_task(
            email_service.send_contact_notification,
            contact_data.name,
            contact_data.email,
            contact_data.phone or "",
            contact_data.message
        )
        
        return await db.contact_submissions.find_one({"id": contact_id}, {"_id": 0})
    
    @router.get("/", response_model=List[ContactSubmission])
    async def list_contacts(
        unread_only: bool = False,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """List all contact submissions (admin only)"""
        await verify_admin(credentials, db)
        
        query = {"is_read": False} if unread_only else {}
        contacts = await db.contact_submissions.find(query, {"_id": 0}).sort("created_at", -1).to_list(500)
        return contacts
    
    @router.get("/{contact_id}", response_model=ContactSubmission)
    async def get_contact(
        contact_id: str,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """Get a specific contact submission (admin only)"""
        await verify_admin(credentials, db)
        
        contact = await db.contact_submissions.find_one({"id": contact_id}, {"_id": 0})
        if not contact:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Contact not found")
        return contact
    
    @router.put("/{contact_id}/read")
    async def mark_as_read(
        contact_id: str,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """Mark contact as read (admin only)"""
        await verify_admin(credentials, db)
        
        contact = await db.contact_submissions.find_one({"id": contact_id})
        if not contact:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Contact not found")
        
        await db.contact_submissions.update_one({"id": contact_id}, {"$set": {"is_read": True}})
        return {"message": "Marked as read"}
    
    @router.put("/{contact_id}/notes")
    async def update_notes(
        contact_id: str,
        notes: str,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """Update contact notes (admin only)"""
        await verify_admin(credentials, db)
        
        contact = await db.contact_submissions.find_one({"id": contact_id})
        if not contact:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Contact not found")
        
        await db.contact_submissions.update_one({"id": contact_id}, {"$set": {"notes": notes}})
        return await db.contact_submissions.find_one({"id": contact_id}, {"_id": 0})
    
    @router.delete("/{contact_id}", status_code=status.HTTP_204_NO_CONTENT)
    async def delete_contact(
        contact_id: str,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """Delete a contact submission (admin only)"""
        await verify_admin(credentials, db)
        
        contact = await db.contact_submissions.find_one({"id": contact_id})
        if not contact:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Contact not found")
        
        await db.contact_submissions.delete_one({"id": contact_id})
        logger.info(f"Deleted contact: {contact_id}")
    
    return router
